core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala

```diff
@@ -103,6 +103,14 @@ private[spark] class MetricsSystem private (
     sinks.foreach(_.start)
   }
 
+  // Same as start but this method only registers sinks
```

Comment:
I see it is a duplication of the start() method without the registerSources() but what about extending the start() method with a new boolean flag (like registerStaticSources:Boolean = true) and using this flag to decide registerSources() should be called or not. Then in AM you can call it with false.

Reply:
Thanks @attilapiros for looking at this. I agree with your proposal. I'll provide an update to the PR.

Reply:
I was thinking a bit more about this PR and in case of client mode it could make sense to have some of this static metrics as the driver and AM are separated. What about (assuming the boolean flag is there) calling the start method with false only in case of cluster mode?

Reply:
I would split the case in 2 topics: 
1. I belive that CodeGenerator and HiveExternalCatalog metrics don't make sense in the context of AM, so they can be safely removed. 
2. The JVM metrics may be relevant as you mentioned. Although in the current version I see the problem that the JVM metrics for the AM appear without any application id nor prefix, so they are difficult to process. I guess this part can be improved if we think JVM metrics for AM can be of interest.

Reply:
I can just fully agree with you.

Reply:
Just to close the loop here: I like the extra flag to `start` instead of the separate method.

As for AM vs. driver, the way I understand things, in cluster mode, there will be two `MetricSystem` instances - one for the AM and one for the driver, so you shouldn't lose any driver metrics.

JVM metrics for the client-mode AM can be interesting, but that's generally not a source of problems that I've noticed, so we can probably punt on it for now.

Reply:
I have implemented the extra flag to start and tested the patch on a YARN cluster using a graphite sink for the metrics.